src/Microsoft.AspNetCore.Sockets.Http/Transports/WebSocketsTransport.cs

```diff
@@ -126,69 +117,93 @@ public WebSocketsTransport(WebSocketOptions options, IChannelConnection<Message>
+                if (incomingMessage.Count > 1)
+                {
+                    var messageBuffer = new byte[totalBytes];
+                    var offset = 0;
+                    for (var i = 0; i < incomingMessage.Count; i++)
+                    {
+                        Buffer.BlockCopy(incomingMessage[i].Array, 0, messageBuffer, offset, incomingMessage[i].Count);
+                        offset += incomingMessage[i].Count;
+                    }
+
+                    message = new Message(messageBuffer, messageType, receiveResult.EndOfMessage);
+                }
+                else
+                {
+                    var buffer = new byte[incomingMessage[0].Count];
+                    Buffer.BlockCopy(incomingMessage[0].Array, incomingMessage[0].Offset, buffer, 0, incomingMessage[0].Count);
+                    message = new Message(buffer, messageType, receiveResult.EndOfMessage);
+                }
+
+                _logger.LogInformation("Passing message to application. Payload size: {0}", message.Payload.Length);
+                while (await _application.Output.WaitToWriteAsync())
+                {
+                    if (_application.Output.TryWrite(message))
+                    {
+                        incomingMessage.Clear();
+                        break;
+                    }
+                }
             }
         }
 
-        private async Task StartSending(IWebSocketConnection ws)
+        private async Task StartSending(WebSocket ws)
         {
             while (await _application.Input.WaitToReadAsync())
             {
                 // Get a frame from the application
-                Message message;
-                while (_application.Input.TryRead(out message))
+                while (_application.Input.TryRead(out var message))
                 {
                     if (message.Payload.Length > 0)
                     {
                         try
                         {
-                            var opcode = message.Type == MessageType.Binary ?
-                                WebSocketOpcode.Binary :
-                                WebSocketOpcode.Text;
-
-                            var frame = new WebSocketFrame(
-                                endOfMessage: message.EndOfMessage,
-                                opcode: _lastFrameIncomplete ? WebSocketOpcode.Continuation : opcode,
-                                payload: ReadableBuffer.Create(message.Payload));
+                            var messageType = message.Type == MessageType.Binary ?
+                                WebSocketMessageType.Binary :
+                                WebSocketMessageType.Text;
 
-                            _lastFrameIncomplete = !message.EndOfMessage;
+                            if (_logger.IsEnabled(LogLevel.Debug))
+                            {
+                                _logger.LogDebug(
```

Comment:
Use the logging syntax that maintains structured log messages: `_logger.LogDebug("Sending frame: Fin={fin}, Payload={length} bytes", message.EndOfMessage, message.Payload.Length);`

I wish we could some how tap in to `FormattedString` to make this work while still preserving structured logging :)